README.md
# IPK Project 2 ZETA: Network sniffer
**Title**: IPK ZETA: Network sniffer\
**Author**: Elena Marochkina\
**Language**: en-US\
**Supported OS**: Linux\
**Gitea**: [xmaroc00_sniffer](https://git.fit.vutbr.cz/xmaroc00/IPK_Project2_2023)

# Table of Contents
1. [Description](#description)
    1. [Compilation](#compilation)
    2. [Usage](#usage)
    3. [Program Details](#program-details)
2. [Testing](#testing)
3. [License](#license)
4. [Bibliography](#bibliography)

## Description

This is a program that uses the pcap library to capture and analyze network traffic (**a packet sniffer**). 

The program can:
- print the names of available network interfaces
- capture packets based on various criteria, such as protocol type and port number
- print information about captured packets, such as their source and destination 
IP and mac addresses, port numbers, length, time in RFC 3339 format and payload data. 

## Compilation
### Linux

To compile the program, you can run **make build**, and to remove the ipk-sniffer executable
file, you can run **make clean**.

### Windows
The program can't be compiled for Windows.

## Usage
The client application requires  arguments to run:
- **-i eth0** (just one interface to sniff) or **--interface**. If this parameter is not specified (and any other parameters as well), or if only -i/--interface is specified without a value (and any other parameters are unspecified), a list of active interfaces is printed .
- **-t** or **--tcp** (displays TCP segments and is optionally complemented by -p functionality).
- **-u** or **--udp** (displays UDP datagrams and is optionally complemented by-p functionality).
- **-p 23** (extends previous two parameters to filter TCP/UDP based on port number; if this parameter is not present, then no filtering by port number occurs; if the parameter is given, the given port can occur in both the source and destination part of TCP/UDP headers).
- **--icmp4** (displays only ICMPv4 packets).
- **--icmp6** (displays only ICMPv6 echo request/response).
- **--arp** (display only ARP frames).
- **--ndp** (displays only ICMPv6 NDP packets).
- **--igmp** (displays only IGMP packets).
- **--mld** (displays only MLD packets).
Unless protocols are explicitly specified, all (i.e., all content, regardless of protocol) are considered for printing.
- **-n 10** (specifies the number of packets to display, i.e., the "time" the program runs; if not specified, consider displaying only one packet, i.e., as if -n 1)
All arguments can be in any order.

      Usage: ./ipk-sniffer [-i interface | --interface interface] {-p port [--tcp|-t] [--udp|-u]} [--arp] [--icmp4] [--icmp6] [--igmp] [--mld] {-n num}

## Program Details
### Included libraries
The code uses the following libraries:

- **stdio.h**: Provides basic input/output operations such as printf and scanf.
- **string.h**: Provides functions for manipulating strings such as strcpy and strcat.
- **getopt.h**: Provides functions for parsing command-line arguments.
- **stdbool.h**: Defines the Boolean data type and its possible values.
- **stdlib.h**: Provides functions for memory allocation and manipulation such as malloc and free.
- **ctype.h**: Provides functions for character handling such as isdigit.
- **pcap.h**: Provides functions for capturing and processing network packets.
- **time.h**: Provides functions for working with date and time.
- **net/ethernet.h**: Defines Ethernet packet structures.
- **arpa/inet.h**: Provides functions for working with Internet addresses.
- **sys/socket.h**: Provides functions for working with sockets.

### Global Variables
**err_buf[PCAP_ERRBUF_SIZE]** - err_buf is a character array (i.e., a string) of size
PCAP_ERRBUF_SIZE that is used to store error messages generated by libpcap, a 
library for capturing network traffic.

### Program Components
#### Sniffer

The **"sniffer.h"** contains multiple helper functions to print information about 
packets captured by a sniffer program.

**Functions**

**void timeval_to_string(struct timeval time, const char \*format)**

*Input parameters*:

- **time**: a struct timeval representing the timestamp of the captured packet.
- **format**: a constant character array representing the format of the timestamp string.

*Output*: none

*Description*: This function converts a struct timeval timestamp to a string format.

**print_ip_and_ports(u_char *packet)**

*Input parameters*:
- **packet**: u_char pointer representing the captured packet.

*Output*: none

*Description*: This function prints the source and destination IP addresses and ports of the captured packet.
- If the protocol of the packet is TCP or UDP, the function extracts the source and destination ports from the packet, and prints them out as well. The function uses bitwise operations to extract the 16-bit port numbers from the packet. The offset values for the source and destination ports depend on whether the packet is IPv4 or IPv6.
- If the Ethernet type of the packet is ARP (Address Resolution Protocol), the function extracts the source and destination IP addresses from the packet and prints them out, without considering any port numbers.

**print_package_data(uint8_t \*data, size_t size)**

*Input Parameters*:
- **data**: A pointer to an array of bytes (uint8_t) that contains the package data.
- **size**: The size of the array (data) in bytes.

*Output*:
This function does not return any value. Instead, it prints the package data in a formatted manner.

*Description*:
The function takes an array of bytes and its size as input and prints the data in a
formatted way to the console, displaying the hexadecimal representation of each byte
as well as its ASCII representation. 

**packet_handler(u_char \*args, const struct pcap_pkthdr \*header, const u_char \*packet)**

*Input Parameters*:
- **args**: A pointer to the arguments passed to the function.
- **header**: A pointer to the pcap_pkthdr structure that contains the packet header.
- **packet**: A pointer to the packet data.

*Output*:
  This function does not return any value. Instead, it prints various information about the packet in a formatted 
manner.

*Description*:
- The function first prints the timestamp of the packet in a formatted manner. 
- Then, it prints the source and destination MAC addresses of the packet. 
- After that, it prints the length of the packet. 
- The function then calls the print_ip_and_port function to print the IP address 
and port of the packet. 
- Finally, the function prints the package data in a formatted manner using the print_package_data function.

#### Filter

The **"filter.h"** header provides functions for creating packet filters using the libpcap library. 
These functions allow the user to specify various criteria, such as protocol type,
port number, and protocol flags, to filter the packets that will be captured.

**Functions**

**void add_protocol(char\* filter_str, const char\* protocol_name, bool* empty)**

*Input parameters*:
- **filter_str**: a character array representing the filter expression to which the protocol name will be added.
- **protocol_name**: a constant character array representing the name of the protocol to be added to the filter expression.
- **empty**: a boolean pointer indicating whether the filter expression is empty or not.

*Output*: none

*Description*: 
This function is a helper function to add a protocol to the filter expression.

**void add_port(char\* filter_str, int port_num, bool\* empty, int tcp_flag, int udp_flag)**

*Input parameters*:
- **filter_str**: a character array representing the filter expression to which the port number will be added.
- **port_num**: an integer representing the port number to be added to the filter expression.
- **empty**: a boolean pointer indicating whether the filter expression is empty or not.
- **tcp_flag**: an integer indicating whether the TCP flag is set or not.
- **udp_flag**: an integer indicating whether the UDP flag is set or not.

*Output*: none

*Description*: This function is a helper function to add a port to the filter expression.

**void create_filter_expression(char\* filter_str, int port, int tcp_flag, int udp_flag, int arp_flag, int icmp4_flag, int icmp6_flag, int ndp_flag, int igmp_flag, int mld_flag)**

*Input parameters:*
- **filter_str**: a character array representing the filter expression to be created.
- **port**: an integer representing the port number to be added to the filter expression. It can be -1 if the port flag is not set.
- **tcp_flag**: an integer indicating whether the TCP flag is set or not.
- **udp_flag**: an integer indicating whether the UDP flag is set or not.
- **arp_flag**: an integer indicating whether the ARP flag is set or not.
- **icmp4_flag**: an integer indicating whether the ICMPv4 flag is set or not.
- **icmp6_flag**: an integer indicating whether the ICMPv6 flag is set or not.
- **ndp_flag**: an integer indicating whether the NDP flag is set or not.
- **igmp_flag**: an integer indicating whether the IGMP flag is set or not.
- **mld_flag**: an integer indicating whether the MLD flag is set or not.

*Output*: none

*Description*: This function creates a filter expression based on the given flags.

#### Main 
The main function for the sniffer program implements the core logic for sniffing network packets on a specified network
interface. 

It begins by declaring and initializing several variables with default values, including the network interface,
port, packet count, and various protocol flags.

The function then defines the available command line options and their corresponding parameters using the *getopt_long* 
function. It loops through all the command line options and their parameters to parse and store the user-specified options.

If no interface is provided, the function calls the **print_interfaces** function to print the names of all available 
network interfaces and exits the program.

If an interface is provided, the function uses the **pcap_lookupdev** function to look up the interface's network device,
opens the device using the **pcap_open_live** function, and sets a capture filter using the **pcap_compile** and **pcap_setfilter** 
functions based on the specified protocol flags.

Finally, the function enters a loop to capture packets using the **pcap_loop function**. 
For each packet, it passes the packet to the packet_handler function, which parses and prints the packet's details to 
the console. If a packet count was specified, the loop exits after capturing the specified number of packets. 

Once all packets have been captured, the function closes the capture device using the **pcap_close**
function and exits the program.

## Testing
1. The scenario where no interface is provided.

**Input:**
```
make && sudo ./ipk-sniffer
make && sudo ./ipk-sniffer -i
make && sudo ./ipk-sniffer --interface
```

**Output:**

```
en0
awdl0
en3
en4
stf0
```

2. The scenario where protocols are given but no interface is provided.

**Input:**
```
make && sudo ./ipk-sniffer -t
make && sudo ./ipk-sniffer --udp
make && sudo ./ipk-sniffer --udp -t -p 23 --arp
etc
```

**Output:**
```
Error: no interface provided
```

4. The scenario where interface is provided but it doesn't exist.

**Input:**
```
make && sudo ./ipk-sniffer -i 8908  
make && sudo ./ipk-sniffer --interface 89fhjhfbw
```
**Output:**
```
Error: 8908: No such device exists (BIOCSETIF failed: Device not configured)
```

5. The scenario where port number is not valid.

**Input:**
```
make && sudo ./ipk-sniffer -i en0 -p 65536
```

**Output:**
```
Invalid port number. Please enter a port number less than 65535
```

6. The scenario where protocols and interface are given; the number of packets is set by 1 or not set.
Port number is valid.

**Input:**
```
make && sudo ./ipk-sniffer -i en0 -p 443 --tcp -u -n 1
make && sudo ./ipk-sniffer -i en0 -p 443 --tcp --udp
```

**Output:**
```
timestamp: 2023-04-17T00:43:30.772+02:00
src mac: b2:67:b5:a7:7e:64
dst mac: d0:88:0c:77:22:81
frame length: 70 bytes
src IP: 149.154.167.151
dst IP: 172.20.10.5
src port: 443
dst port: 65123

0x0000: d0 88 0c 77 22 81 b2 67 b5 a7 7e 64 08 00 45 02 ...w"..g..~d..E.
0x0010: 00 38 d4 a9 00 00 35 06 bd c9 95 9a a7 97 ac 14 .8....5.........
0x0020: 0a 05 01 bb fe 63 5a 07 5b 23 5d d0 6e cb 80 18 .....cZ.[#].n...
0x0030: 80 00 81 ad 00 00 01 01 08 0a cc 4e 24 f1 8c 87 ...........N$...
0x0040: c8 3d 90 4b 29 82                               .=.K).
```

7. The scenario where interface and port is given and tcp flag is configured (*port is printed*)

**Input:**
```
make && sudo ./ipk-sniffer -i en0 --tcp -p 443
make && sudo ./ipk-sniffer -i en0 -t -p 443
make && sudo ./ipk-sniffer -interface en0 --tcp -p 443
make && sudo ./ipk-sniffer -interface en0 -t -p 443
```
**Output:**
```
timestamp: 2023-04-17T16:24:04.423+02:00
src mac: c0:c9:e3:89:20:8a
dst mac: d0:88:0c:77:22:81
frame length: 66 bytes
src IP: 45.14.133.24
dst IP: 192.168.0.102
src port: 443
dst port: 60260

0x0000: d0 88 0c 77 22 81 c0 c9 e3 89 20 8a 08 00 45 00 ...w"..... ...E.
0x0010: 00 34 10 26 40 00 34 06 c3 69 2d 0e 85 18 c0 a8 .4.&@.4..i-.....
0x0020: 00 66 01 bb eb 64 77 a9 f7 e7 4a 1c ab 80 80 10 .f...dw...J.....
0x0030: 02 60 f6 af 00 00 01 01 08 0a ac 2f 3c 5b cd 15 .`........./<[..
0x0040: 02 8a                                           ..
```

8. The scenario where interface is given and arp flag is configured (*no port is printed*)
 The same will work for icmp4, icmp6, igmp, mld protocols.

**Input:**
```
make && sudo ./ipk-sniffer -i en0 --arp 
```

**Output:**
```
timestamp: 2023-04-17T16:27:34.361+02:00
src mac: c0:c9:e3:89:20:8a
dst mac: ff:ff:ff:ff:ff:ff
frame length: 42 bytes
src IP: 192.168.0.1
dst IP: 192.168.0.101

0x0000: ff ff ff ff ff ff c0 c9 e3 89 20 8a 08 06 00 01 .......... .....
0x0010: 08 00 06 04 00 01 c0 c9 e3 89 20 8a c0 a8 00 01 .......... .....
0x0020: 00 00 00 00 00 00 c0 a8 00 65                   .........e
```

**All received data were verified using the program Wireshark!**

## License
This project is licensed under the GNU General Public License - see the LICENSE file for details.

## Bibliography
1. TCPDUMP/LIBPCAP public repository: http://www.tcpdump.org/
3. Using libpcap in C: https://www.devdungeon.com/content/using-libpcap-c
4. RFC 792 - Internet Control Message Protocol a RFC 4443 - ICMPv6 
5. RFC 826 - ARP
6. RFC 3339 -  Date and Time on the Internet: Timestamps


CHANGELOG.md
# IPK Project 2 ZETA: Network sniffer
**Title**: IPK ZETA: Network sniffer\
**Author**: Elena Marochkina\
**Language**: en-US\
**Supported OS**: Linux\
**Gitea**: [xmaroc00_sniffer](https://git.fit.vutbr.cz/xmaroc00/IPK_Project2_2023)

# Brief Description
The program is designed to capture and analyze network traffic that is transmitted over a network interface. 

The code contains the main function and several helper functions for initializing and setting up the program's various 
options, including interface selection, filtering by protocol, and setting a capture limit. 
The code uses the libpcap library to capture network packets and apply filters based on user-defined parameters. 

The program provides a command-line  interface with options for selecting a network interface, specifying the type 
of traffic to capture, setting a packet capture limit, and filtering by protocol. 

The program prints the captured packets to the standard output in a human-readable format.

# Known Limitations
The program appears to have a few limitations, some of which are:

1. The code does not have input validation or error handling for all possible user inputs. For example, if a user 
provides an invalid interface name or filter expression, the program will terminate abruptly without providing any 
feedback on the error.
2. The code only supports a limited number of protocol filters (TCP, UDP, ARP, ICMP4, ICMP6, NDP, IGMP, MLD) and port 
filtering. It does not support filtering based on other protocol headers or packet contents.
3. The program does not have any options for output formatting, such as saving the output to a file.
4. The program does not have any options for capturing packets from a remote machine or monitoring multiple interfaces
simultaneously.

filter.c
#include "filter.h"

void add_protocol(char* filter_str, const char* protocol_name, bool* empty) {
    if (*empty) {
        strcat(filter_str, protocol_name);
        *empty = false;
    } else {
        strcat(filter_str, " or ");
        strcat(filter_str, protocol_name);
    }
}

void add_port(char* filter_str, int port_num, bool* empty, int tcp_flag, int udp_flag) {
    char port_number[20];
    sprintf(port_number, "%d", port_num);

    // Adding port to the expression
    strcpy(filter_str, "port ");
    strcat(filter_str, port_number);
    if (tcp_flag && udp_flag || !tcp_flag && !udp_flag){
        strcat(filter_str, " and (tcp or udp)");
    } else if (tcp_flag) {
        strcat(filter_str, " and tcp ");
    } else if (udp_flag) {
        strcat(filter_str, " and udp ");
    }
    // Set the empty flag to false
    *empty = false;
}

void
create_filter_expression(char* filter_str, int port, int tcp_flag, int udp_flag, int arp_flag, int icmp4_flag, int icmp6_flag, int ndp_flag, int igmp_flag, int mld_flag) {
    bool port_flag = (port != -1);
    bool empty = true;

    // Add the port and tcp/udp flags to the filter expression if the port flag is set
    port_flag? add_port(filter_str, port, &empty, tcp_flag, udp_flag) : 0;

    // Add the protocols to the filter expression if the protocol flags are set
    if (!port_flag){
        tcp_flag? add_protocol(filter_str, "tcp", &empty) : 0;
        udp_flag? add_protocol(filter_str, "udp", &empty) : 0;
    }
    arp_flag? add_protocol(filter_str, "arp", &empty) : 0;
    icmp4_flag? add_protocol(filter_str, "icmp", &empty) : 0;
    icmp6_flag? add_protocol(filter_str, "icmp6 and (icmp6[0] == 128 or icmp6[0] == 129)", &empty) : 0;
    igmp_flag? add_protocol(filter_str, "igmp", &empty) : 0;
    mld_flag? add_protocol(filter_str, "(icmp6 and (icmp6[0] == 135 or icmp6[0] == 136))", &empty) : 0;
    ndp_flag? add_protocol(filter_str, "(icmp6 and (icmp6[0] == 130 or icmp6[0] == 131))", &empty) : 0;
}

sniffer.c
#include "sniffer.h"

void timeval_to_string(struct timeval time, const char *format) {
    time_t time_in_sec = time.tv_sec;
    char buffer[64];
    struct tm* timeInfo;
    time_t unix_time = (time_t) time_in_sec;
    timeInfo = localtime(&unix_time);
    strftime(buffer, 64, "%Y-%m-%dT%H:%M:%S", timeInfo);

    // Get the local time zone
    char timezone_buffer[6];
    strftime(timezone_buffer, 6, "%z", timeInfo);

    // Convert the timezone offset to ISO 8601 format
    char timezone_offset[7];
    timezone_offset[0] = timezone_buffer[0];
    timezone_offset[1] = timezone_buffer[1];
    timezone_offset[2] = timezone_buffer[2];
    timezone_offset[3] = ':';
    timezone_offset[4] = timezone_buffer[3];
    timezone_offset[5] = timezone_buffer[4];
    timezone_offset[6] = '\0';

    printf("%s.%03ld%s\n", buffer, time.tv_usec / 1000, timezone_offset);
}

void print_ip_and_ports(u_char *packet) {
    // Print out the source IP address for v4 and v6
    const struct ethernet *ethernet = (struct ethernet*)(packet);

// Check if the Ethernet type is IPv4
    if (ntohs(ethernet->ether_type) == ETHERTYPE_IP) {
        // Define char arrays to store the source and destination IP addresses
        char src_ip[INET_ADDRSTRLEN];
        char dst_ip[INET_ADDRSTRLEN];
        // Extract the source IP address from the packet using inet_ntop and store it in src_ip
        inet_ntop(AF_INET, packet + 26, src_ip, INET_ADDRSTRLEN);
        // Extract the destination IP address from the packet using inet_ntop and store it in dst_ip
        inet_ntop(AF_INET, packet + 30, dst_ip, INET_ADDRSTRLEN);
        // Print out the source and destination IP addresses
        printf("src IP: %s\n", src_ip);
        printf("dst IP: %s\n", dst_ip);

        // Print out the source and destination ports for TCP and UDP
        // Check if the protocol is TCP
        if (packet[23] == 6 || packet[23] == 17) {
            // Extract the source port from the packet
            uint16_t src_port = (packet[34] << 8) | packet[35];
            // Extract the destination port from the packet
            uint16_t dst_port = (packet[36] << 8) | packet[37];
            // Print out the source and destination ports
            printf("src port: %d\n", src_port);
            printf("dst port: %d\n", dst_port);
        }
    }
// Check if the Ethernet type is IPv6
    else if (ntohs(ethernet->ether_type) == ETHERTYPE_IPV6) {
        // Define a pointer to an IPv6 header and set it to the start of the IPv6 packet
        struct ip6_hdr *ip6 = (struct ip6_hdr*)(packet + sizeof (struct ethernet));
        // Define char arrays to store the source and destination IPv6 addresses
        char src_ip[INET6_ADDRSTRLEN];
        char dst_ip[INET6_ADDRSTRLEN];
        // Extract the source IPv6 address from the packet using inet_ntop and store it in src_ip
        inet_ntop(AF_INET6, &ip6->ip6_src, src_ip, INET6_ADDRSTRLEN);
        // Extract the destination IPv6 address from the packet using inet_ntop and store it in dst_ip
        inet_ntop(AF_INET6, &ip6->ip6_dst, dst_ip, INET6_ADDRSTRLEN);
        // Print out the source and destination IPv6 addresses
        printf("src IP: %s\n", src_ip);
        printf("dst IP: %s\n", dst_ip);

        // Print out the source and destination ports for TCP and UDP
        // Check if the protocol is TCP
        if (packet[20] == 6 || packet[20] == 17) {
            // Extract the source port from the packet
            uint16_t src_port = (packet[54] << 8) | packet[55];
            // Extract the destination port from the packet
            uint16_t dst_port = (packet[56] << 8) | packet[57];
            // Print out the source and destination ports
            printf("src port: %d\n", src_port);
            printf("dst port: %d\n", dst_port);
        }
    }
    else if (ntohs(ethernet->ether_type) == ETHERTYPE_ARP) {
        char src_ip[INET_ADDRSTRLEN];
        char dst_ip[INET_ADDRSTRLEN];
        inet_ntop(AF_INET, packet + 28, src_ip, INET_ADDRSTRLEN);
        inet_ntop(AF_INET, packet + 38, dst_ip, INET_ADDRSTRLEN);
        printf("src IP: %s\n", src_ip);
        printf("dst IP: %s\n", dst_ip);
    }
}

void print_package_data(uint8_t *data, size_t size) {
    // Initialize two size_t variables, i and j
    size_t i, j;
    // Loop through the array in increments of 16 (i.e. print 16 bytes at a time)
    for (i = 0; i < size; i += 16) {
        // Print the current index in the array (in hexadecimal format)
        printf("0x%04x: ", (unsigned int)i);
        for (j = i; j < i + 16 && j < size; j++) {
            printf("%02x ", data[j]);
        }
        // If there are less than 16 bytes left to print, fill the remaining space with blank spaces (for formatting purposes).
        for (; j < i + 16; j++) {
            printf("   ");
        }
        // Print a space before printing the ASCII representation of the bytes.
        // Loop through the next 16 bytes, or until the end of the array (whichever comes first).
        for (j = i; j < i + 16 && j < size; j++) {
            if (data[j] >= 32 && data[j] <= 126) {
                printf("%c", data[j]);
            } else {
                printf(".");
            }
        }
        // Print a newline character at the end of each 16-byte row.
        printf("\n");
    }
    printf("\n\n");
}

void packet_handler(u_char *args, const struct pcap_pkthdr *header, const u_char *packet) {
    // print the timestamp of the packet
    printf("timestamp: ");
    timeval_to_string(header->ts, "%Y-%m-%d %H:%M:%S");

    // print mac addresses
    printf("src mac: %02x:%02x:%02x:%02x:%02x:%02x\n", packet[6], packet[7], packet[8], packet[9], packet[10], packet[11]);
    printf("dst mac: %02x:%02x:%02x:%02x:%02x:%02x\n", packet[0], packet[1], packet[2], packet[3], packet[4], packet[5]);

    // print the length of the packet
    printf("frame length: %d bytes\n", header->len);

    // print IP address and port

    print_ip_and_ports((u_char*) packet);

    // print the package data
    printf("\n");
    print_package_data((uint8_t*) packet, header->len);
}

sniffer.h
#ifndef IPK_PROJECT2_2023_SNIFFER_H
#define IPK_PROJECT2_2023_SNIFFER_H

#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <stdlib.h>
#include <ctype.h>
#include <pcap.h>
#include <time.h>
#include <net/ethernet.h>
#include <arpa/inet.h>
#include <sys/socket.h>
#include <netinet/ip6.h>

// Structure of an internet header, naked of options
struct ethernet {
    u_char ether_dhost[ETHER_ADDR_LEN]; // Destination host address
    u_char ether_shost[ETHER_ADDR_LEN]; // Source host address
    u_short ether_type;
};

// This function takes in a timeval struct and a format string and prints out the time in the specified format
void timeval_to_string(struct timeval time, const char *format);

// This function takes in a pointer to an array of bytes (u_char) and prints out the source and destination IP addresses
void print_ip(u_char *packet);

// This function takes in a pointer to an array of bytes (uint8_t), as well as the size of the array
void print_package_data(uint8_t *data, size_t size);

// This function is called by pcap_loop for every packet that is captured
void packet_handler(u_char *args, const struct pcap_pkthdr *header, const u_char *packet);

#endif //IPK_PROJECT2_2023_SNIFFER_H


main.c
/**
 * IPK Project 2 ZETA: Network sniffer.
 * @author
 *  xmaroc00, Elena Marochkina

 * @date   17.04.2023
 */

#include "main.h"
#include "filter.h"
#include "sniffer.h"

// This function prints the names of all available network interfaces
void print_interfaces() {
    pcap_if_t *interfaces;
    // pcap_findalldevs returns a linked list of available interfaces
    if (pcap_findalldevs(&interfaces, err_buf) == 0) {
        printf("\n");
        // loop through all interfaces and print their names
        for (pcap_if_t *interface = interfaces; interface != NULL; interface = interface->next) {
            printf("%s\n", interface->name);
        }
        printf("\n");
        // free the memory allocated by pcap_findalldevs
        pcap_freealldevs(interfaces);
    } else {
        // if pcap_findalldevs fails, print an error message
        fprintf(stderr, "Error in pcap_findall_devs(): %s\n", err_buf);
    }
};

int main(int argc, char *argv[]) {
    // initialize variables with default values
    char interface[255] = "";
    int port = -1;
    bool tcp_flag = 0;
    bool udp_flag = 0;
    bool icmp4_flag = 0;
    bool icmp6_flag = 0;
    bool arp_flag = 0;
    bool ndp_flag = 0;
    bool igmp_flag = 0;
    bool mld_flag = 0;
    int count = -1;

    pcap_t *handle;
    struct bpf_program filter;
    char filter_exp[255] = "";

    // define the program's command line options and their corresponding parameters
    static struct option long_options[] =
            {
                    {"interface", optional_argument, 0, 'i'},
                    {"tcp",       no_argument,       0, 't'},
                    {"udp",       no_argument,       0, 'u'},
                    {"arp",       no_argument,       0, 1},
                    {"icmp4",     no_argument,       0, 2},
                    {"icmp6",     no_argument,       0, 3},
                    {"ndp",       no_argument,       0, 4},
                    {"igmp",      no_argument,       0, 5},
                    {"mld",       no_argument,       0, 6},
                    {0,           0,                 0, 0}
            };

    int option_index = 0;
    int option;

    int interface_flag = 0;

    // loop through all the command line options and their parameters using getopt_long
    while ((option = getopt_long(argc, argv, "i::p:tun:", long_options, &option_index)) != -1) {
        switch (option) {
            case 0:
                break;
            case 'i':
                // Check if interface name is provided
                if (argv[optind] == NULL) {
                    interface_flag = 1;
                    break;
                }
                strncpy(interface, argv[optind], 254);
                interface_flag = 1;
                break;
            case 'p':
                // Check if port value is a number
                for (size_t i = 0; i < optarg[i] != '\0'; i++)
                    if (!isdigit(optarg[i])) {
                        fprintf(stderr, "Error: port value is not digit\n");
                        exit(1);
                    }
                // convert the port value to an integer
                port = atoi(optarg);
                if (port > 65535) {
                    fprintf(stderr, "Invalid port number. Please enter a port number less than 65535\n");
                    exit(EXIT_FAILURE);
                }
                break;
            case 't':
                tcp_flag = 1;
                break;
            case 'u':
                udp_flag = 1;
                break;
            case 'n':
                // Check if count value is a number
                for (size_t i = 0; i < optarg[i] != '\0'; i++) {
                    if (!isdigit(optarg[i])) {
                        fprintf(stderr, "Error: number of packets is not digit\n");
                        exit(1);
                    }
                }
                // Convert the count value to an integer
                count = atoi(optarg);
                break;
            case 1:
                arp_flag = 1;
                break;
            case 2:
                icmp4_flag = 1;
                break;
            case 3:
                icmp6_flag = 1;
                break;
            case 4:
                ndp_flag = 1;
                break;
            case 5:
                igmp_flag = 1;
                break;
            case 6:
                mld_flag = 1;
                break;
            case '?':
                // Print an error message for unknown arguments and exit the program
                fprintf(stderr, "Unknown argument %s\n", argv[optind - 1]);
                exit(1);
            default:
                // Print the correct usage for the program and exit
                fprintf(stderr,
                        "Usage: %s [-i interface | --interface interface] {-p port} {[--tcp|-t] [--udp|-u] [--arp] [--icmp] } {-n num}\n",
                        argv[0]);
                exit(1);
        }
    }

    // If no interface is provided, print a list of available interfaces and exit
    if (strcmp(interface, "") == 0) {
        if ((interface_flag == 1 && argc == 2) || (interface_flag == 0 && argc == 1)) {
            print_interfaces();
            exit(0);
        } else {
            fprintf(stderr, "Error: no interface provided\n");
            exit(1);
        }
    }

    // If the number of packets to capture is not provided, set it to 1
    if (count == -1) {
        count = 1;
    }

    // Check if the interface is valid
    handle = pcap_open_live(interface, BUFSIZ, 1, 1000, err_buf);
    if (handle == NULL) {
        fprintf(stderr, "Error: %s\n", err_buf);
        exit(2);
    }

    // Check if the interface provides Ethernet headers
    if (pcap_datalink(handle) != DLT_EN10MB) {
        fprintf(stderr, "Device %s doesn't provide Ethernet headers - not supported\n", interface);
        return (2);
    }

    // Create the filter expression
    create_filter_expression(filter_exp, port, tcp_flag, udp_flag, arp_flag, icmp4_flag, icmp6_flag, ndp_flag,
                             igmp_flag, mld_flag);

    // Compile the filter expression
    if (pcap_compile(handle, &filter, filter_exp, 0, PCAP_NETMASK_UNKNOWN) == -1) {
        fprintf(stderr, "Couldn't parse filter %s: %s\n", filter_exp, pcap_geterr(handle));
        return (2);
    }

    // Apply the filter expression
    if (pcap_setfilter(handle, &filter) == -1) {
        fprintf(stderr, "Couldn't install filter %s: %s\n", filter_exp, pcap_geterr(handle));
        return (2);
    }

    // Capture packets
    pcap_loop(handle, count, packet_handler, NULL);
    // Free the filter expression
    pcap_freecode(&filter);
    // Close the session
    pcap_close(handle);

    return 0;
}